testy/runtime.py

```diff
@@ -1,25 +1,103 @@
-import threading
+import multiprocessing
 import time
-import studentCode
+import sys
+import traceback
+
+import stateManager
+import studentAPI
+
+from runtimeUtil import *
+
+
+# TODO:
+# 0. Set up testing code for the following features.
+# DONE 1. Have student code go through api to modify state.
+# DONE 2. Imposing timeouts on student code (infinite loop, try-catch)
+# 3. Figure out how to kill student thread.
+# 4. Integrate with Bob's socket code: spin up a communication process
+# 5. stateManager throw badThing on processNameNotFound
+# 6. refactor process startup code: higher order function
+# 7. Writeup how all this works
+# 8. Investigate making BadThing extend exception
+
+allProcesses = {}
+badThings = multiprocessing.Condition()
+globalBadThing = "unititialized globalBadThing"
 
-STUDENT_THREAD_NAME = "student_thread"
 
 def runtime():
-  state = [0]
-  runStudentCode(state)
-  while True:
-    time.sleep(1)
-
-def runStudentCode(state):
-  studentThread = threading.Thread(target=runStudentCodeHelper, name=STUDENT_THREAD_NAME, args=(state,))
-  studentThread.daemon = True
-  studentThread.start()
-
-def runStudentCodeHelper(state):
-  nextCall = time.time()
-  while True:
-    studentCode.main(state)
-    nextCall += 1
-    time.sleep(nextCall - time.time())
+  badThingsQueue = multiprocessing.Queue()
+  stateQueue = multiprocessing.Queue()
+  spawnProcess = processFactory(badThingsQueue, stateQueue)
+  restartCount = 0
+  try:
+    spawnProcess(PROCESS_NAMES.STATE_MANAGER, startStateManager)
+    while True:
+      if restartCount >= 5:
+        print(RUNTIME_INFO.DEBUG_DELIMITER_STRING.value)
+        print("Too many restarts, terminating")
+        break
+      print(RUNTIME_INFO.DEBUG_DELIMITER_STRING.value)
+      print("Starting studentCode attempt: %s" % (restartCount,))
+      spawnProcess(PROCESS_NAMES.STUDENT_CODE, runStudentCode)
+      while True:
+        globalBadThing = badThingsQueue.get(block=True)
+        print(RUNTIME_INFO.DEBUG_DELIMITER_STRING.value)
+        print(globalBadThing)
+        if (globalBadThing.event == BAD_EVENTS.STUDENT_CODE_ERROR) or \
+            (globalBadThing.event == BAD_EVENTS.STUDENT_CODE_TIMEOUT):
+          break
+      stateQueue.put([SM_COMMANDS.RESET])
+      restartCount += 1
+  except:
+    print(RUNTIME_INFO.DEBUG_DELIMITER_STRING.value)
+    print("Funtime Runtime Had Too Much Fun")
+    print(traceback.print_exception(*sys.exc_info()))
+
+def runStudentCode(badThingsQueue, stateQueue, pipe):
+  try:
+    import signal
+    def timed_out_handler(signum, frame):
+      raise TimeoutError("studentCode timed out")
+    signal.signal(signal.SIGALRM, timed_out_handler)
+
+    import studentCode
```

Comment:
Importing arbitrary student code is dangerous because it [executes it](http://stackoverflow.com/questions/6523791/why-is-python-running-my-module-when-i-import-it-and-how-do-i-stop-it)